File: 2020/3/part2.py
from functools import reduce

rows = []
with open('input.txt', 'r') as f:
    rows = list(map(str.strip, f.readlines()))

slopes = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]

ans = []

for slope in slopes:
    trees = 0
    pos = [0, 0]

    while pos[0] < len(rows) - 1:
        pos[1] %= len(rows[0])

        if rows[pos[0]][pos[1]] == '#':
            trees += 1

        pos[0] += slope[1]
        pos[1] += slope[0]

    ans.append(trees)

print(ans)

res = reduce(lambda x, y: x * y, ans)
print(res)

File: 2020/3/cpp/part2.cpp
#include <iostream>
#include <vector>

using namespace std;

int main()
{
    vector<int> xs = {1, 3, 5, 7, 1};
    vector<int> ys = {1, 1, 1, 1, 2};
    vector<string> map;

    string s;
    while (getline(cin, s))
    {
        map.push_back(s);
    }

    long long res = 1;
    for (int i = 0; i < xs.size(); i++)
    {
        int x = 0;
        int y = 0;
        int trees = 0;

        while (y < map.size())
        {
            if (map[y][x] == '#')
            {
                trees++;
            }

            x = (x + xs[i]) % map[y].size();
            y = (y + ys[i]);
        }

        res *= trees;
    }

    cout << res << endl;
}